src/networkd.c

```diff
@@ -1466,6 +1457,60 @@ _netplan_netdef_write_networkd(
     return TRUE;
 }
 
+gboolean
+_netplan_networkd_write_wait_online(const NetplanState* np_state, const char* rootdir)
+{
+    // Hash set of non-optional interfaces to wait for
+    GHashTable* non_optional_interfaces = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    NetplanStateIterator iter;
+    netplan_state_iterator_init(np_state, &iter);
+    while (netplan_state_iterator_has_next(&iter)) {
+        const NetplanNetDefinition* def = netplan_state_iterator_next(&iter);
+        if (def->backend != NETPLAN_BACKEND_NETWORKD)
+            continue;
+        /* When activation-mode is used we default to being optional.
+         * Otherwise, systemd might wait indefinitely for the interface to
+         * come online.
+         */
+        if (!(def->optional || def->activation_mode)) {
+            if (!netplan_netdef_has_match(def)) { // no matching => single interface
+                g_hash_table_add(non_optional_interfaces, g_strdup(def->id));
+            } else if (def->set_name) { // matching on a single interface
+                g_hash_table_add(non_optional_interfaces, g_strdup(def->set_name));
+            } else { // matching on potentially multiple interfaces
+                // XXX: we shouldn't run this enumeration for every NetDef...
+                _netplan_enumerate_interfaces(def, non_optional_interfaces, rootdir);
+            }
+        }
+    }
+
+    // create run/systemd/system/systemd-networkd-wait-online.service.d/
+    const char* override = "/run/systemd/system/systemd-networkd-wait-online.service.d/10-netplan.conf";
+    // The "ConditionPathIsSymbolicLink" is Netplan's s-n-wait-online enablement symlink,
+    // as we want to run -wait-only only if enabled by Netplan.
+    GString* content = g_string_new("[Unit]\n"
+        "ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait-online.service\n");
+    if (g_hash_table_size(non_optional_interfaces) == 0) {
+        _netplan_g_string_free_to_file(content, rootdir, override, NULL);
+        g_hash_table_destroy(non_optional_interfaces);
+        return FALSE;
+    }
+
+    // We have non-optional interface, so let's wait for those explicitly
+    GHashTableIter idx;
+    gpointer ifname;
+    g_string_append(content, "\n[Service]\nExecStart=\n"
```

Comment:
Why do we need this empty `ExecStart=`?

Reply:
It is valid to have multiple `ExecStart=`. Using an empty one clears the current list so that you can completely override it.